src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoCommon.cs

```diff
@@ -83,48 +85,49 @@ public override JsonConverter ConverterBase
             }
         }
 
-        public override object GetValueAsObject(object obj)
+        public override object? GetValueAsObject(object? obj)
         {
             if (IsPropertyPolicy)
             {
                 return obj;
             }
 
             Debug.Assert(HasGetter);
-            return Get(obj);
+            return Get!(obj);
         }
 
-        public override void SetValueAsObject(object obj, object value)
+        public override void SetValueAsObject(object? obj, object? value)
         {
             Debug.Assert(HasSetter);
-            TDeclaredProperty typedValue = (TDeclaredProperty)value;
+            TDeclaredProperty typedValue = (TDeclaredProperty)value!;
 
             if (typedValue != null || !IgnoreNullValues)
             {
-                Set(obj, typedValue);
+                Set!(obj, typedValue);
             }
         }
 
-        private JsonPropertyInfo _elementPropertyInfo;
+        private JsonPropertyInfo? _elementPropertyInfo;
 
         private void SetPropertyInfoForObjectElement()
         {
+            Debug.Assert(ElementClassInfo != null);
             if (_elementPropertyInfo == null && ElementClassInfo.PolicyProperty == null)
             {
                 _elementPropertyInfo = ElementClassInfo.CreateRootProperty(Options);
             }
         }
 
-        public override bool TryCreateEnumerableAddMethod(object target, out object addMethodDelegate)
+        public override bool TryCreateEnumerableAddMethod(object target, [NotNullWhen(true)] out object? addMethodDelegate)
         {
             SetPropertyInfoForObjectElement();
-            Debug.Assert((_elementPropertyInfo ?? ElementClassInfo.PolicyProperty) != null);
+            Debug.Assert((_elementPropertyInfo ?? ElementClassInfo!.PolicyProperty) != null);
 
-            addMethodDelegate = (_elementPropertyInfo ?? ElementClassInfo.PolicyProperty).CreateEnumerableAddMethod(RuntimeClassInfo.AddItemToObject, target);
+            addMethodDelegate = (_elementPropertyInfo ?? ElementClassInfo!.PolicyProperty!).CreateEnumerableAddMethod(RuntimeClassInfo.AddItemToObject!, target);
```

Comment:
@steveharter, @layomia - please review this. Can `AddItemToObject` ever be null here?

Reply:
In this case, `AddItemToObject` can't be null since the caller checks for it (in `ReadStackFrame`). Using `!` makes sense.